addons/stock_picking_wave/models/stock_picking_wave.py

```diff
@@ -0,0 +1,255 @@
+        for wave in self:
+            wave.move_ids = wave.move_line_ids.move_id
+
+    def _set_move_ids(self):
+        # Hack to be able to have the 'delete' button on stock move
+        pass
+
+    @api.onchange('move_ids')
+    def _onchange_move_ids(self):
+        deleted_moves = self._origin.move_ids - self.move_ids._origin
+        if deleted_moves:
+            self.move_line_ids = self.move_line_ids.filtered(lambda l: l.move_id != deleted_moves)
+
+    @api.onchange('picking_ids')
+    def _onchange_picking_ids(self):
+        deleted_pickings = self._origin.picking_ids - self.picking_ids._origin
+        if deleted_pickings:
+            self.move_line_ids = self.move_line_ids.filtered(lambda l: l.picking_id != deleted_pickings)
+
+    # -------------------------------------------------------------------------
+    # CRUD
+    # -------------------------------------------------------------------------
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get('name', '/') == '/':
+                vals['name'] = self.env['ir.sequence'].next_by_code('picking.wave') or 'New'
+        return super().create(vals_list)
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('picking_ids'):
+            wave_without_picking_type = self.filtered(lambda wave: not wave.picking_type_id)
+            if wave_without_picking_type:
+                picking = self.picking_ids and self.picking_ids[0]
+                wave_without_picking_type.picking_type_id = picking.picking_type_id.id
+        if vals.get('picking_type_id'):
+            self._sanity_check()
+        return res
+
+    # -------------------------------------------------------------------------
+    # Action methods
+    # -------------------------------------------------------------------------
+    def action_cancel(self):
+        self.ensure_one()
+        self.state = 'cancel'
+        return True
+
+    def action_done(self):
+        self._check_company()
+        if not self.env.context.get('button_validate_wave_ids'):
+            self = self.with_context(button_validate_wave_ids=self.ids)
+        res = self._pre_action_done_hook()
+        if res is not True:
+            return res
+        new_move_vals = []
+        for move in self.move_line_ids.move_id:
+            if move.move_line_ids <= self.move_line_ids:
+                continue
+            qty = sum([line.qty_done for line in move.move_line_ids & self.move_line_ids])
+            qty_split = move.product_uom._compute_quantity(move.product_uom_qty - qty, move.product_id.uom_id, rounding_method='HALF-UP')
+            new_move = move._split(qty_split, waves=self)
+            new_move_vals += new_move
+        if new_move_vals:
+            new_moves = self.env['stock.move'].create(new_move_vals)
+            new_moves._action_confirm(merge=False)
+        self.move_line_ids.move_id._action_done()
```

Comment:
Don't forget to:
trigger_assign on validate move

+ _action_done on stock.picking if last line was done in the wave